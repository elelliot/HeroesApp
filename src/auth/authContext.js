import { createContext } from "react";


// es un comp por eso es upperCase
export const AuthContext = createContext()