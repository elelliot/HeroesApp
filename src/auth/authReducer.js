import { types } from "../components/types/types";

// const state = {
    //     name: 'Elliot',
    //     logged: true
    // }
    
    
    
//Autenticamos el login
//El action modifica el state, y cuando eso pase, react dispara los cambios y redibuja
export const authReducer = (state = {}, action ) => {
    
    // console.log(action)
    switch( action.type ) {
        case types.login:
            return {
                ...action.payload, //así me evito poner 1 por uno ej: action.payload.name (ver el types.js)
                logged: true,
            }

        case types.logout:
            return {
                logged: false //sin payload
            }

            default:
                return state; //regresamos el state vacío
    }
}