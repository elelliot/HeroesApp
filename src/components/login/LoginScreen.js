import { useContext } from "react";
import { useNavigate } from "react-router-dom";
import { AuthContext } from "../../auth/authContext";

import { types } from "../types/types";

export const LoginScreen = () => {
    


    //para mandar la accion
    const { dispatch }  = useContext( AuthContext );

    //useNavigate es una función que permite navegar a otras pantallas (Router-Dom V6)
    const navigate = useNavigate();

    const handleLogin = () => {


        //autenticando login y mandando la acción
        const action = {
            type: types.login,
            payload: { name:"Elliot Miranda" }//se tiene que ver en la navbar
        }

        dispatch(action);


        /*ahora que tenemos la última ruta guardada en lastPath (la guardamos al autenticarnos en PrivateRoute)
        , la usamos al loguearnos para que nos llevé allá */
        
        const lastPath = localStorage.getItem('lastPath') || '/marvel';

        /*navegamos a lastPath cuando demos click en login y el "replace" omite el login(o la ágina anterior) 
        del historial en cuanto le de al botón, osea, no podré volver con el botón de "atrás" */
        navigate( lastPath, {
            replace: true
        });
        
    }

    return (
        <div className="container mt-5">
            <h1>Login</h1>
            <hr />

            <button
                className="btn btn-primary"
                onClick = { handleLogin }
            >
                Login
            </button>
        </div>
    )
}