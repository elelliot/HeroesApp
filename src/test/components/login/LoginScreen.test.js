import { mount } from "enzyme";
import { MemoryRouter,Routes, Route } from "react-router-dom";
import { AuthContext } from "../../../auth/authContext";
import { LoginScreen } from "../../../components/login/LoginScreen";
import { types } from "../../../components/types/types";

//Como se usa el useNavigate, debo poner esto y luego lo probaremos
const mockNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate,
}));


describe('Pruebas en <LoginScreen />', () => {
    

    //Creamos contexto para mandarlo
    const contextValue = {
        dispatch: jest.fn(),
        user:{
            logged: false
        }
    }


    //Mount global
    const wrapper = mount( 
        <AuthContext.Provider value={ contextValue }>
            <MemoryRouter initialEntries={ ['/login'] }>
                <Routes>
                    <Route path="/login" element={ <LoginScreen /> } />                                
                </Routes>
            </MemoryRouter>
        </AuthContext.Provider>    
    );


    test('Debe hacer match con Snapshot', () => {
        expect( wrapper ).toMatchSnapshot();
    });



    test('Debe realizar el dispatch y la navegación', () => {
        //Login con el click...
        const handleClick = wrapper.find('button').prop('onClick');
        handleClick();

        //-Que el dispatch se ejecute con la acción, {name: Elliot}
        expect( contextValue.dispatch ).toHaveBeenCalledWith({
            type: types.login,
            payload: { name:"Elliot Miranda" }
        });


        //que el mockNavigate sea llamado con (/marvel (si no hay una por default, será esa), {replace:true})
        expect( mockNavigate ).toHaveBeenCalledWith( '/marvel', { replace: true } )


        //"Simulamos" la visita de una nueva pagina (Ya no ocupamos hacer mock)
        localStorage.setItem('lastPath', '/dc');

        //Volvemos a usar el Click del login...
        handleClick();
        //... y al volver a loguearnos este debería ser el valor con el que se llamó mockNavigate
        expect( mockNavigate ).toHaveBeenCalledWith( '/dc', { replace: true } )
    })
    
    
})
