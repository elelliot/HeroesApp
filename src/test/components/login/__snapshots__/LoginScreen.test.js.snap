// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <LoginScreen /> Debe hacer match con Snapshot 1`] = `
<div
  className="container mt-5"
>
  <h1>
    Login
  </h1>
  <hr />
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Login
  </button>
</div>
`;
